### DISCUSSION TITLE
negative value for variables %timer1% to %timer16% added in #14619

### DISCUSSION THREAD

**Discussion by rapi3** [arendst/Tasmota#15205] [Q&A] [answered]
Hi,
I noticed that value for defined timer2 -00:30 = 750 instead of expected -30
and defined timer3 -01:00 = 780 instead of expected -60
is this normal ??

this mean that negative numbers are calculated as 720+(positive value of timer) and it complicate a lot the use of %timer% in rule if it have a negative value:
for example I can't use it as generic if timer is negative:
%sunset%+%timer2%
%sunset%-%timer2%
%sunrise%-%timer3%
%sunrise%+%timer3%

```
22:18:25.348 CMD: timer1
22:18:25.355 MQT: s/P05/RESULT = {"Timer1":{"Enable":1,"Mode":1,"Time":"00:30","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":3}}
22:18:26.588 CMD: timer2
22:18:26.595 MQT: s/P05/RESULT = {"Timer2":{"Enable":1,"Mode":2,"Time":"-00:30","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":3}}
22:18:32.784 CMD: timer3
22:18:32.792 MQT: s/P05/RESULT = {"Timer3":{"Enable":1,"Mode":1,"Time":"-01:00","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":3}}
22:18:49.836 CMD: timer4
22:18:49.843 MQT: s/P05/RESULT = {"Timer4":{"Enable":1,"Mode":2,"Time":"04:00","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":3}}
```
this rule use generic to check the values:
```
ON Time#Initialized DO backlog var3=%sunrise%+%timer1%; var4=%sunset%-%timer2%; var5=%sunrise%-%timer3%; var6=%sunset%+%timer4%; var7=%sunset%; var8=%sunrise%; var9=%time%; var11=%timer1%; var12=%timer2%; var13=%timer3%; var14=%timer4%; event checktime=%time% ENDON
```
result of variables 4 and 5 not good, it will be out of time range:
```
22:21:06.403 CMD: var
22:21:06.411 MQT: s/P05/RESULT = {"Var1":"1","Var2":"0","Var3":"425.000","Var4":"377.000","Var5":"-385.000","Var6":"1367.000","Var7":"1127.000","Var8":"395.000","Var9":"1324.000","Var10":"","Var11":"30.000","Var12":"750.000","Var13":"780.000","Var14":"240.000","Var15":"","Var16":""}
```
this rule is changed to have a good result but is no good for generic:
```
ON Time#Initialized DO backlog var3=%sunrise%+%timer1%; var4=%sunset%-30; var5=%sunrise%-60; var6=%sunset%+%timer4%; var7=%sunset%; var8=%sunrise%; var9=%time%; var11=%timer1%; var12=%timer2%; var13=%timer3%; var14=%timer4%; event checktime=%time% ENDON
```
result of variables 4 and 5 it is correct:
```
22:31:08.413 CMD: var
22:31:08.421 MQT: s/P05/RESULT = {"Var1":"1","Var2":"0","Var3":"425.000","Var4":"1097.000","Var5":"335.000","Var6":"1367.000","Var7":"1127.000","Var8":"395.000","Var9":"1331.000","Var10":"","Var11":"30.000","Var12":"750.000","Var13":"780.000","Var14":"240.000","Var15":"","Var16":""}
```

**edit:**
_it would be nice if the value of timer will be stored calculated as defined by user: time/sunset/sunrise +/-value_

**Comment by barbudor** [comment]
Can't reproduce your problem
```
22:43:21.613 CMD: backlog rule3; timer1; event test
22:43:21.707 MQT: stat/nodemcu/RULE = {"Rule3":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":47,"Free":464,"Rules":"on event#test do var9 = %sunset%+%timer1% endon"}}
22:43:21.919 MQT: stat/nodemcu/TIMER = {"Timer1":{"Enable":1,"Mode":2,"Time":"01:00","Window":0,"Days":"1111111","Repeat":0,"Output":1,"Action":0}}
22:43:22.168 MQT: stat/nodemcu/EVENT = {"Event":"Done"}
22:43:22.185 RUL: EVENT#TEST performs "var9 = 1143+60"
22:43:22.193 MQT: stat/nodemcu/VAR = {"Var9":"1203.000"}
```
Sunset is 19:03 = 1143 , timer1 is 01:00 = 60 => sunset+timer1 = 1143+60 = 1203

**Reply by rapi3** [reply]
try with sunset/sunrise - timer: -01:00
![Screenshot_2022-03-21_21-02-41](https://user-images.githubusercontent.com/18531150/159369434-cbb04dc8-bd46-4a21-bfe5-9e78a6417ece.png)

**Reply by barbudor** [reply]
So checking in the code, it seems that negative values are stored as 12 + abs(value)
So -00:30 is stored as (12+0) * 60 + 30 = 750
And -01:00 is stored as (12+1) * 60 + 0 = 780

I understand that it could makes sense that for modes 1 and 2, the value of %timerX% reports a correct negative interger and not vertabim how it is stored internally.

@alexasf would you like to update your code ?
You can see the logic here : https://github.com/arendst/Tasmota/blob/b5e96117ba91eb4243cc4e5d5967da96d7cfc189/tasmota/xdrv_09_timers.ino#L319-L324

Or I can have a look if you are not available.

**Reply by rapi3** [reply]
@barbudor
please have a look when you have time if possible

also it would be super if you can improve/fix:
so the value of timer will be stored calculated as defined by user: time/sunset/sunrise +/-value and not only if it is a fixed value because it limit the use case when user define timer for sunrise/sunset+offset.

thank you very much

**Reply by barbudor** [reply]
> so the value of timer will be stored calculated as defined by user: time/sunset/sunrise +/-value and not only if it is a fixed value because it limit the use case when user define timer for sunrise/sunset+offset.

IMHO this is quite questionable
The value returned would be valid only for short time (for today) and not representative of the actual configuration.
As you can always do as you already did by adding/substracting `%sunset%` and `%sunrise%` to achieve this.

I'll try to look at this during the week-end

**Reply by alexasf** [reply]
> So checking in the code, it seems that negative values are stored as 12 + abs(value) So -00:30 is stored as (12+0) * 60 + 30 = 750 And -01:00 is stored as (12+1) * 60 + 0 = 780
> 
> I understand that it could makes sense that for modes 1 and 2, the value of %timerX% reports a correct negative interger and not vertabim how it is stored internally.
> 
> @alexasf would you like to update your code ? You can see the logic here :
> 
> https://github.com/arendst/Tasmota/blob/b5e96117ba91eb4243cc4e5d5967da96d7cfc189/tasmota/xdrv_09_timers.ino#L319-L324
> 
> Or I can have a look if you are not available.

Yes, it can be released. But I'm currently busy with another project. And at the moment I don't have a module to test. And on working devices, I will not debug the code. I'll try to implement it as soon as I get a chance.

**Reply by barbudor** [reply]
No problem @alexasf 
I've done it, hoe you don't mind
PR to come:
```
22:23:25.617 CMD: backlog rule3;timer1;timer2;timer3;timer4;event test
22:23:25.707 MQT: stat/nodemcu/RULE = {"Rule3":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":267,"Free":244,"Rules":"ON event#test DO backlog var3=%sunrise%+(%timer1%); var4=%sunset%+(%timer2%); var5=%sunrise%+(%timer3%); var6=%sunset%+(%timer4%); var7=%sunset%; var8=%sunrise%; var9=%time%; var11=%timer1%; var12=%timer2%; var13=%timer3%; var14=%timer4%; event checktime=%time% ENDON"}}
22:23:25.950 MQT: stat/nodemcu/TIMER = {"Timer1":{"Enable":1,"Mode":1,"Time":"00:30","Window":0,"Days":"1111111","Repeat":1,"Action":0}}
22:23:26.198 MQT: stat/nodemcu/TIMER = {"Timer2":{"Enable":1,"Mode":2,"Time":"-00:30","Window":0,"Days":"1111111","Repeat":1,"Action":0}}
22:23:26.451 MQT: stat/nodemcu/TIMER = {"Timer3":{"Enable":1,"Mode":1,"Time":"-01:00","Window":0,"Days":"1111111","Repeat":1,"Action":0}}
22:23:26.702 MQT: stat/nodemcu/TIMER = {"Timer4":{"Enable":1,"Mode":2,"Time":"04:00","Window":0,"Days":"1111111","Repeat":1,"Action":0}}
22:23:26.951 MQT: stat/nodemcu/EVENT = {"Event":"Done"}
22:23:26.982 RUL: EVENT#TEST performs "backlog var3=402+(30); var4=1149+(-30); var5=402+(-60); var6=1149+(240); var7=1149; var8=402; var9=1343; var11=30; var12=-30; var13=-60; var14=240; event checktime=1343"
22:23:27.000 MQT: stat/nodemcu/VAR = {"Var3":"432.000"}
22:23:27.254 MQT: stat/nodemcu/VAR = {"Var4":"1119.000"}
22:23:27.502 MQT: stat/nodemcu/VAR = {"Var5":"342.000"}
22:23:27.706 MQT: stat/nodemcu/VAR = {"Var6":"1389.000"}
22:23:27.907 MQT: stat/nodemcu/VAR = {"Var7":"1149.000"}
22:23:28.113 MQT: stat/nodemcu/VAR = {"Var8":"402.000"}
22:23:28.364 MQT: stat/nodemcu/VAR = {"Var9":"1343.000"}
22:23:28.610 MQT: stat/nodemcu/VAR = {"Var11":"30.000"}
22:23:28.861 MQT: stat/nodemcu/VAR = {"Var12":"-30.000"}
22:23:29.112 MQT: stat/nodemcu/VAR = {"Var13":"-60.000"}
22:23:29.361 MQT: stat/nodemcu/VAR = {"Var14":"240.000"}
```

@rapi3 please check also the fixed rule as now `%TIMERx%` return a negative value when needed, you need to *add* to `%sunrise%`/`%sunset%`

**Reply by barbudor** [reply]
https://github.com/arendst/Tasmota/pull/15224

**Reply by rapi3** [reply]
Thank you very much !

**Comment by barbudor** [comment, marked as answer]
I'm thinking about the absolute time but that's not obvious

What value would you expect `%TIMERx%` to return for results that fall the previous day or the next day ? 
Depending on daylight saving and season:
- SUNRISE - 07:00 could be negative 
- SUNSET + 07:00 could be over 24:00

**Reply by rapi3** [reply]
_I am expecting the value returned for %timer% to use the same value as for sunset/sunrise +/- offset calculated in that moment or after time initialized._

In my case I use %timer% variable to check at boot time after time initialized if switches and rules has to be ON or OFF as I defined timers.
After boot finished the switches and rules will be controlled as usual by defined timers.

The case you list - spot on - are more some particular case ( user logic error or the moment when timezone change the time ) and you can return the value limited to 0-24 interval as normal:
if sunrise - offset < 0h then return sunrise or 0
if sunset + offset > 24h then return sunset or 24

**Reply by barbudor** [reply]
So I decided to remain coherent with Mode 0 and return a time-of-today between 0 (00:00) and 1439 (23:59)
My current sunrise is 06:40 (400) and sunset is 19:11 (1151) :
```
14:13:56.914 CMD: backlog rule3;timers;event test
14:13:56.992 MQT: stat/nodemcu/RULE = {"Rule3":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":204,"Free":307,"Rules":"ON event#test DO backlog var1 %timer1%;var2 %timer2%; var3 %timer3%;var4 %timer4%;var5 %timer5%;var6 %timer6%;var7 %timer7%;var8 %timer8%;var9 %timer9%;var10 %timer10%;var11 %sunrise%;var12 %sunset% endon"}}
14:13:57.219 MQT: stat/nodemcu/TIMERS = {"Timers":"ON",
  "Timer1":{"Enable":1,"Mode":1,"Time":"00:10","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer2":{"Enable":1,"Mode":1,"Time":"-00:10","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer3":{"Enable":1,"Mode":1,"Time":"11:59","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer4":{"Enable":1,"Mode":1,"Time":"-11:59","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer5":{"Enable":1,"Mode":2,"Time":"00:10","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer6":{"Enable":1,"Mode":2,"Time":"-00:10","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer7":{"Enable":1,"Mode":2,"Time":"11:59","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer8":{"Enable":1,"Mode":2,"Time":"-11:59","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer9":{"Enable":1,"Mode":1,"Time":"00:00","Window":0,"Days":"1111111","Repeat":1,"Action":3},
  "Timer10":{"Enable":1,"Mode":2,"Time":"00:00","Window":0,"Days":"1111111","Repeat":1,"Action":3},
...
}

14:13:57.508 MQT: stat/nodemcu/VAR = {"Var1":"410"}       => 06:40 + 00:10 => 06:50 = 410
14:13:57.758 MQT: stat/nodemcu/VAR = {"Var2":"390"}       => 06:40 - 00:10 => 06:30 = 390
14:13:58.009 MQT: stat/nodemcu/VAR = {"Var3":"1119"}      => 06:40 + 11:59 => 18:39 = 1119
14:13:58.259 MQT: stat/nodemcu/VAR = {"Var4":"1121"}      => 06:40 - 11:59 => 18:41 = 1121
14:13:58.510 MQT: stat/nodemcu/VAR = {"Var5":"1161"}      => 19:11 + 00:10 => 19:21 = 1161
14:13:58.760 MQT: stat/nodemcu/VAR = {"Var6":"1141"}      => 19:11 - 00:10 => 19:01 = 1141
14:13:59.010 MQT: stat/nodemcu/VAR = {"Var7":"430"}       => 19:11 + 11:59 => 07:10 = 430
14:13:59.259 MQT: stat/nodemcu/VAR = {"Var8":"432"}       => 19:11 - 11:59 => 07:12 = 432
14:13:59.460 MQT: stat/nodemcu/VAR = {"Var9":"400"}       => 06:40 + 00:00 => 06:40 = 400
14:13:59.710 MQT: stat/nodemcu/VAR = {"Var10":"1151"}     => 19:11 + 00:00 => 19:11 = 1151
14:13:59.961 MQT: stat/nodemcu/VAR = {"Var11":"400"}      => sunrise
14:14:00.211 MQT: stat/nodemcu/VAR = {"Var12":"1151"}     => sunset
```

I think this remains coherent with @alexasf original PR to allow calculating if %time% is before or after a Timer event time (not counting the random option)

**Reply by rapi3** [reply]
Super.
Many thanks for your work !

**Reply by barbudor** [reply]
New PR https://github.com/arendst/Tasmota/pull/15230
